### PR TITLE
IdModel: loop promotion cleanup

### PR DESCRIPTION
**Pull request by naoyam**
No logic change. Mostly mechanical cleanup.

Replaced the test-specific IdModel subclass with a callback interface. The callback interface allows to save all necessary temporary results for validation. No more duplication of `buildLoopPromotionMap`. (Related comment: https://github.com/NVIDIA/Fuser/pull/2220#discussion_r1596296138)

To introduce the callback interface, moved the loop promotion part out of `IdModel` to its own builder class.

### CODE REVIEW

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

File: csrc/id_model/loop_promotion.cpp
Comment (naoyam) [inline csrc/id_model/loop_promotion.cpp:1]: Almost exact copy from id_model.cpp

File: csrc/id_model/id_model.cpp [hunk 1]
```diff
@@ -582,263 +591,6 @@ void IdModel::buildLoopGraph() {
   idGraph(IdMappingMode::LOOP).validateConsistency();
 }
 
-std::unordered_map<ValGroup, IterDomain*> IdModel::buildLoopPromotionMap(
```
Comment (naoyam) [inline csrc/id_model/id_model.cpp:585]: Moved to loop_promotion.cpp

**Comment by naoyam** [comment]
CI passed at 2b605cd

**Comment by naoyam** [comment]
!build

**Review by zasdfgbnm — APPROVED** [review]
Is the callback only used in tests? Or it will be used somewhere else in the future?

**Comment by naoyam** [comment]
Only the tests. I don't have any planned use case either.

**Pull request merged**